### ISSUE TITLE
Intellisense auto popup

### ISSUE THREAD

**Issue by vasily-kirichenko** [Microsoft/visualfsharp#406]
I'd like intellisence to auto popup as I type an identifier, not by pressing `Ctrl+Space`. It's how code completion works in just every IDE and editor. Are there any practical reasons to not auto popup intellisense? Maybe it's too slow? I think it's a very simple and straightforward fix.

(newbies even think there's no intellisense in F# editor at all, see https://social.msdn.microsoft.com/Forums/en-US/3f37c607-ea62-4e3f-b38a-102d37bb756f/newbie-question-intelisense-does-not-work-at-all?forum=fsharpgeneral)

/cc @latkin @dsyme @KevinRansom  

@dungpa we could theoretically emulate this in Power Tools by sending `Ctrl+Space` to the editor, but it feels as a very dirty hack.

**Comment by isaacabraham** [comment]
@vasily-kirichenko I think that this is standard Visual Studio behaviour though - do you really want to have different behaviour when you switch between languages?

**Comment by vasily-kirichenko** [comment]
No, it's different from C# editor behavior. In C# intellisense always auto pops up:

![6677](https://cloud.githubusercontent.com/assets/873919/7441726/952de23a-f0fd-11e4-8d37-fb1a39d5f56b.gif)

In F# it auto pops up after the dot only (i.e. for type members only):

![6677](https://cloud.githubusercontent.com/assets/873919/7441739/f603ad9c-f0fd-11e4-89c3-c4b8c2f3a98b.gif)

(here I had to press `Ctrl+Space` to launch intellisense twice: for `Console` and for `System`). 
So now it seems even easier to fix since it already works ok after dots.

**Comment by braden** [comment]
Are you sure that is not a Resharper feature? I don't have auto-completion like that in C#, although I might have something disabled.

**Comment by forki** [comment]
I don't have resharper installed (and I'm sure @vasily-kirichenko is also avoiding it) and for me C# intellisense is working as in the small screencast.

**Comment by vasily-kirichenko** [comment]
Yes, I've not used ReSharper for almost 3 years.

**Comment by vasily-kirichenko** [comment]
I think adding `case VsCommands2K.TYPECHAR:` here https://github.com/Microsoft/visualfsharp/blob/1cae3585b1d1a7fecba7262d7436523b42f805bd/vsintegration/src/vs/FsPkgs/FSharp.LanguageService/FSharp.LanguageService.Base/ViewFilter.cs#L533 would fix this issue. I don't have VS 2015 installed to test it.

**Comment by isaacabraham** [comment]
Good point. I was thinking of intellisense after a dot, which as you say is enabled in F#. Something like this sounds like a very low hanging fruit that would definitely aid introduction to f#.

I have 2015 installed, happy to test it out. Especially as in 2015 you lose CTRL + . with power tools!

**Comment by latkin** [comment]
Yeah this bugs me sometimes, too.  This must have been considered in the past, I'm guessing it was just judged to be too slow.

**Comment by forki** [comment]
Is it possible to make a switch and disable if needed.

**Comment by enricosada** [comment]
I really like vspt, with features toggles

**Comment by vasily-kirichenko** [comment]
Yes, it's a great idea to add a switch to disable it.

**Comment by dsyme** [comment]
I don't think this has really been tried - in particular since VS2012 the completion code is now properly asynchronous in all code paths, so populating immediately is much more reasonable.  IIRC http://tryfsharp.org does this and performance seems fine.

So perhaps we should just give it a go.  It would be great if this could somehow be trialled in VFPT though.

**Comment by vasily-kirichenko** [comment]
If we're gonna trigger it from VFPT, this notes may be useful https://github.com/jaredpar/VsVim/blob/master/Src/VsVim/KeyboardInputRouting.txt

**Comment by enricosada** [comment]
maybe is better inside visual f# with a feature flag in tools menu? 
the code is there, is easier to enable than from outside, and little features are nice (and up-for-grab ready)

**Comment by dsyme** [comment]
Because VF# is ultimately an enterprise-supported product it doesn't really feel the right place for flighting features. VFPT gets us to flight the feature properly ( VFPT can deliver even to VS2012)

FWIW it may be plausible for VFPT to replace the intellisense implementation wholesale (based on FCS though, so the underlying logic is still the same), though they've avoided doing that so far (for good reasons)

**Comment by dungpa** [comment]
Intellisense is a tricky component to extend, let alone replace.
If we add Intellisense support to VFPT, it is most likely to appear as a tab next to VF# Intellisense window. This will create a UI mess without further benefits. That's the reason why we didn't do Intellisense in VFPT.

I think the change is small enough to be done as a PR here. Just need someone brave enough to try it out.

**Comment by isaacabraham** [comment]
Intellisense maybe not the best one - but I personally like the idea of flighting features in VFPT and then seeing them enter into VS proper at a future release. This is exactly the process that the standard VSPT follows.

**Comment by enricosada** [comment]
Intellisense is inside this repo, is easier do that **here + feature flag**. 
The flag can be disabled by default, so **opt-in**
That doesn't mean the feature should be partial or broken.

@dsyme about features/entrerprisey:

Microsoft Azure got :

- **General Availability (GA)**: Features under this category can be used for production purposes. As features under General Availability is fully supported by Microsoft Paid Support Plans.
- **Preview**: Feature categorized under Preview should be used only for testing and development purposes. These features are not ready for production use. A feature with a Preview Status is typically supported through the Microsoft Communities
- **Private Preview**: This status is not very common and you need to provide brief description before you can get access

Feature flags are an easy way to move forward, without break compatibiliy to current users.
Some feature cannot be done like that, some can (like this one), the other way is to hacky.

**Comment by isaacabraham** [comment]
IMHO it's important that intellisense support follows the standard look and feel of other languages in VS. Look at VS2015 where intellisense has been slightly revamped and now contains a little icon and name for the symbol type etc. - this is a good opportunity to try to bring it in line with that as well, ideally,

**Comment by samikrc** [comment]
Any update on this? I have VS 2015 Community with Update 1 and things still work as shown in the screencast.

**Comment by vasily-kirichenko** [comment]
:+1:

**Comment by gomassaerima** [comment]
:+1::+1:

**Comment by VSDekar** [comment]
Any update on this?

**Comment by vasily-kirichenko** [comment]
See https://github.com/Microsoft/visualfsharp/pull/1842 

Correct status of this:

* UI is never, ever blocked on completion
* What's left is to formalize the logic of auto triggering, since it was too aggressive in that PR (similar to VS Code, but still more intelligent, like it was not triggered inside strings and excluded code blocks, etc)

**Comment by vasily-kirichenko** [comment]
@VSDekar I've made a new PR, works OK. Please test it and give feedback https://github.com/Microsoft/visualfsharp/pull/2594

**Comment by vasily-kirichenko** [comment]
Implemented long time ago.
